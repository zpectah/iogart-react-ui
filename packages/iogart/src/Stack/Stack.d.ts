import { StackProps } from './types';

declare const Stack: StackProps;

export default Stack;
