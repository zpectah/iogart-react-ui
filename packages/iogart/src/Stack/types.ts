import {
  IogartCommonUi,
  withChildren,
  withStyles,
  flexboxDirection,
  flexboxAlignContent,
  flexboxAlignItems,
  flexboxAlignSelf,
  flexboxJustifyContent,
  flexboxJustifyItems,
  flexboxJustifySelf,
} from '@iogart-react-ui/types';
import { StackElementTypeKeys } from './enums';

export type StackElementType = keyof typeof StackElementTypeKeys;

export interface StackProps extends IogartCommonUi, withChildren, withStyles {
  elementType?: StackElementType;
  spacing?: number;
  direction?: flexboxDirection;
  alignContent?: flexboxAlignContent;
  alignItems?: flexboxAlignItems;
  alignSelf?: flexboxAlignSelf;
  justifyContent?: flexboxJustifyContent;
  justifyItems?: flexboxJustifyItems;
  justifySelf?: flexboxJustifySelf;
}

export type useStackProps = StackProps;
export type useStackReturn = useStackProps;
