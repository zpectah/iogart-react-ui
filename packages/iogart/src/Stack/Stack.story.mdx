import { Meta, Story } from '@storybook/addon-docs';


import Stack from './Stack';
export const Template = (args) => <Stack {...args} />;

<Meta
    title="Components/Layout"
    component={Stack}
/>

# Stack
`@iogart-react-ui/iogart/Stack`

```javascript
import { Stack } from 'iogart-react-ui/iogart';
```

<Story
  name="Stack"
  args={{
    children: (
        <>
            <div>block 1</div>
            <div>block 2</div>
            <div>block 3</div>
        </>
    ),
  }}>
  {Template.bind({})}
</Story>

