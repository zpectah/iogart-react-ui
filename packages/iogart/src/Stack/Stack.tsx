import React, { FC, createElement } from 'react';

import { contentElementTypeKeys } from '@iogart-react-ui/types';
import { StackProps } from './types';
import useStack from './useStack';

const Stack:FC<StackProps> = (props) => {
    const {
        children,
        style,
        component = contentElementTypeKeys.div,
        ...rest
    } = props;

    const updatedProps = useStack({ style, ...rest });

    return createElement(
        component,
        updatedProps,
        children,
    );
};

export default Stack;
