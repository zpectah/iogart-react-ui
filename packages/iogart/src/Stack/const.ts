export const STACK_CLASSNAME = 'Stack';
