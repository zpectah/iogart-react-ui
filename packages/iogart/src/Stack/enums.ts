export enum StackElementTypeKeys {
  article = 'article',
  aside = 'aside',
  div = 'div',
  footer = 'footer',
  header = 'header',
  main = 'main',
  nav = 'nav',
  p = 'p',
  section = 'section',
  span = 'span',
}
