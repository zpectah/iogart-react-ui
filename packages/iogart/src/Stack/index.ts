export { default as useStack } from './useStack';
export { default as Stack } from './Stack';
export * from './const';
export * from './enums';
export * from './types';
