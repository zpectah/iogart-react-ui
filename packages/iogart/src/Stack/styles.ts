import { CSSProperties } from 'react';
import { createUseStyles } from 'react-jss';
import { withIogartThemeProps } from '@iogart-react-ui/types';
import {
  flexboxDirection,
  flexboxAlignContent,
  flexboxAlignItems,
  flexboxAlignSelf,
  flexboxJustifyContent,
  flexboxJustifyItems,
  flexboxJustifySelf,
} from '@iogart-react-ui/types';

interface useStackStylesProps extends withIogartThemeProps {
  spacing?: number;
  direction?: flexboxDirection;
  alignItems?: flexboxAlignItems;
  alignContent?: flexboxAlignContent;
  alignSelf?: flexboxAlignSelf;
  justifyContent?: flexboxJustifyContent;
  justifyItems?: flexboxJustifyItems;
  justifySelf?: flexboxJustifySelf;
}

const useStackStyles = createUseStyles<'root', useStackStylesProps>({
  root: ({ ...rest }) => {
    const styles: CSSProperties = {
      display: 'flex',
      // backgroundColor: theme.palette.primary.main, // Example for theme
    };
    if (rest.spacing) styles.gap = rest.spacing;
    if (rest.direction) styles.flexDirection = rest.direction || 'row';
    if (rest.alignItems) styles.alignItems = rest.alignItems;
    if (rest.alignContent) styles.alignContent = rest.alignContent;
    if (rest.alignSelf) styles.alignSelf = rest.alignSelf;
    if (rest.justifyContent) styles.justifyContent = rest.justifyContent;
    if (rest.justifyItems) styles.justifyItems = rest.justifyItems;
    if (rest.justifySelf) styles.justifySelf = rest.justifySelf;

    return styles;
  },
});

export default useStackStyles;
