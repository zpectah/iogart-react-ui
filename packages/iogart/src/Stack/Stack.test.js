import renderer from 'react-test-renderer';

import Stack from './Stack';

it('Stack - common', () => {

    const component = renderer.create(
        <Stack> <div>A</div><div>B</div><div>C</div> </Stack>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

});
