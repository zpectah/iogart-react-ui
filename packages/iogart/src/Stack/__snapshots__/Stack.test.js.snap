// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack - common 1`] = `
<div
  className="iogart-ui iogart-stack"
  style={
    {
      "alignContent": "initial",
      "alignItems": "initial",
      "alignSelf": "initial",
      "display": "flex",
      "flexDirection": "column",
      "gap": 0,
      "justifyContent": "initial",
      "justifyItems": "initial",
      "justifySelf": "initial",
    }
  }
>
   
  <div>
    A
  </div>
  <div>
    B
  </div>
  <div>
    C
  </div>
   
</div>
`;
